### PR TITLE
Database Migrations Guide

### PR DESCRIPTION
**Pull request by rygr**

### CODE REVIEW

File: database_migrations/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+Best Practices
+===========
+
+### CODE REVIEW
+
+Submit a pull-request and drop in #pull_requests in Slack for a review.  We should aim for three reviews on all migrations.
+
+- [ ] Another developer with domain knowledge
+- [ ] A member of Leviathon (our TeamSnap Architects)
```
Comment (coreycondardo) [inline database_migrations/README.md]: spelling: Leviathan

**Comment by coreycondardo** [comment]
How about posting it in #schema_changes

**Comment by rygr** [comment]
@teamsnap-full-stack-devs @teamsnap-full-stack-devs @m00dawg wanted to get everyone's opinions on two things.

1) Having a section in our guides devoted to database_migrations best practices.  #schema_changes in Slack gets overloaded with conversations and I think some of them can be resolved with a best practices section in our guides and if the best practices are shared in onboarding, new developers can start putting together where and how we do our migrations.  We could also add to the README in ecco, but the thought was a lot of new developers would be exposed to the guides before the migration repo.

2) @m00dawg has some ideas around the use of permissions tables that live off other larger tables.  What do others think about using permissions tables in cases where we want to expand permission style stuff that live off tables such as rosters.  @m00dawg care to elaborate on this one?  Is this something we want to make a habit of or a case by case scenario.  @glanotte wouldn't mind your thoughts on this as well.

**Comment by m00dawg** [comment]
Sure!

Main concern, in brief, is adding too many columns to our larger tables (or smaller tables that may grow very large in the future) since that will cause some technical debt to fix or make further changes more difficult (read: long RSUs). A small lean table that is likely to get accessed very frequently would be better than a large table being accessed frequency as well. You start getting into some optimizations awarded by having lean tables of fixed-lengths (as a permissions table would be) but I'll quickly get excited and ramble on about that :) The biggest is really where we place said technical debt - in already large tables, or in smaller ones we can reference.

But I think even if we can't have one table to rule them all, standardizing on how we approach permissions across tables and apps would be something I'd be very much in favor of.

**Comment by glanotte** [comment]
My one feeling is that the feature needs to be done or close to it before we migrate the database to support it. There have been a few cases where we go to do the implementation based on the database and find that the database update is not what was needed.

**Comment by rygr** [comment]
@glanotte I can update the guides here to recommend waiting to deploy the database until the code has been tested and ready.  However, I think we need to make sure we are getting buy-in from Ops and others on the database changes up-front.  If a project makes it past QA and is ready for deploy and database changes for scaling are brought to light, we risk additional QA and development time.  I have a tendency to tell ops about migrations when I am ready to schedule the deployment of it.  I think it's worth getting buy-in earlier in the process.

Anyone have any opinions on Tim's comments utilizing permission tables to prevent additional updates to some of our larger tables?

**Comment by m00dawg** [comment]
100% agreed with @rygr here. It sucks to be the bad guy to have to :football: a deploy because the DB design needs tweaking. The earlier we get brought into that process, the better it is for everyone I think.

**Comment by glanotte** [comment]
@rygr totally agree and there I think we can differentiate between reviewing and approving a PR. Or I am even OK with the understanding that things just aren't deployed and migrated until they are done.

**Comment by rygr** [comment]
@teamsnap/full-stack-devs anyone else have opinions on having common migration practices as part of our guide and anyone have any opinions on permission style tables, which @m00dawg outlined above.

**Comment by knappe** [comment]
I've been following along and I think putting this in the guides is a great idea.  I can't speak for the permission tables, but I do think that would make sense for some of our larger tables.  I'm just not sure how feasible a single table will be going forward.  Our permissions are kinda all over the place.

**Comment by m00dawg** [comment]
I think them being all over the place is the crux of the point being made. I dunno if we have to fully centralize it but a `permissions` table that describes abilities in some way tied to a pivot table (e.g. `team_permissions`) is something I could see for some grand unification.

The bigger issue I see is having a bunch of columns being created in the large tables for permissions. It's inflexible for devs given having to do RSUs and causes (minor) customer impact while I end up being a bit of a brake in the process since we have to concerned about performance, risk, backups, and deployment time relative to their size.

So if I grand unified permissions thing can't be realized, next best thing might be more isolated permissions tables for the things that need them. So say `team_permissions` that has team_id as the id and then, if you're fancy and willing to forgo some level of understanding, a varbinary that uses bitmasking in the app. Small fixed-length table which is very efficient but has zero level of understanding by looking at the rows in the table.

For what it's worth, MySQL/MariaDB's own permissions tables use a more understandable approach where you have, e.g. the user_id and then a bunch of ENUM columns that describe the state of that particular permission. It's easy to understand, if a tad wasteful (but not relative to say a varchar). MariaDB actually caches these in-memory as well and, to that point, if we split out permissions in this way we could/should do the same thing using your favorite functional key-value store (mine's memcached).

**Comment by chardoublestar** [comment]
ok to close? @rygr 
I think we've moved past this and this is out of date now.

**Comment by coreycondardo** [comment]
@randy-girard HI RANDY!!!!!!

**Pull request closed without merging**